src/main/kotlin/io/github/mojira/arisa/modules/privacy/PrivacyModule.kt

```diff
@@ -0,0 +1,242 @@
+        .filter {
+            it.body?.let(::containsSensitiveData)?.let { matchResult ->
+                logFoundSensitiveData("in comment with ID ${it.id}", matchResult)
+                return@filter true
+            }
+            return@filter false
+        }
+        .map { { it.restrict("${it.body}$commentNote") } }
+        .toList()
+
+    private fun Issue.logFoundSensitiveData(location: String, matchResult: TextRangeLocation) {
+        // Important: Don't log value (i.e. sensitive data) of match result
+        log.info("$key: Found sensitive data $location at ${matchResult.getLocationDescription()}")
+    }
+
+    private fun Issue.hasAnyAttachmentName(name: String) = attachments.any { it.name == name }
+
+    private fun redactAttachments(issue: Issue, attachments: Collection<RedactedAttachment>): Boolean {
+        var redactedAll = true
+        attachments
+            // Group by uploader in case they uploaded multiple attachments at once
+            .groupBy { it.attachment.uploader?.name!! }
+            .forEach { (uploader, userAttachments) ->
+                val fileNames = mutableSetOf<String>()
+                userAttachments.forEach {
+                    val attachment = it.attachment
+                    val tempDir = Files.createTempDirectory("arisa-redaction-upload").toFile()
+                    val fileName = "redacted_${attachment.name}"
+                    val filePath = getSafeChildPath(tempDir, fileName)
+
+                    if (filePath == null || issue.hasAnyAttachmentName(fileName) || !fileNames.add(fileName)) {
+                        redactedAll = false
+                        // Note: Don't log file name to avoid log injection
+                        log.warn("Cannot redact attachment with ID ${attachment.id} of issue ${issue.key}; file name " +
+                            "is malformed or would clash with other attachment")
+                        tempDir.delete()
+                    } else {
+                        log.info("Redacting attachment with ID ${attachment.id} of issue ${issue.key} because it " +
+                            "contains sensitive data")
+                        filePath.writeText(it.redactedContent)
+                        issue.addAttachmentFromFile(filePath) {
+                            // Once uploaded, delete the temp directory containing the attachment
+                            tempDir.deleteRecursively()
+                        }
+                        // Remove the original attachment
+                        attachment.remove()
+                    }
+                }
+
+                if (fileNames.isNotEmpty()) {
+                    // Use postfix line break to prevent bot signature from appearing as part of last list item
+                    val fileNamesString = fileNames.joinToString(separator = "", postfix = "\n") {
+                        // Use link for attachments
+                        "\n- [^${sanitizeCommentArg(it)}]"
```

Comment:
The aggressive sanitizing here might lead to some attachment links not working, in case they include unicode characters. Which will, arguably, probably barely ever happen. So not sure if it's a big deal.

Reply:
I guess we run the file names through a regex firstly though, so this isn't an issue.

Reply:
We can make `sanitizeCommentArg` more lenient (e.g. by allowing certain character categories) in the future if this turns out to be a problem.

> I guess we run the file names through a regex firstly though

Which one do you mean? We only have a list of regex patterns for names of files which definitely contain sensitive data, but we do not filter malformed file names (except to prevent directory traversal).

Reply:
From what I can tell, attachments are only scanned if they match the hs_err_pid regex. I might be wrong though.

Reply:
It has been some time since I wrote that code, but I don't think there are such requirements. `AccessTokenRedactor` only checks for the file content (more specifically for `--accessToken`), and the `PrivacyModuleTest` passes without specifying any special file names.

Reply:
Ah yes you're right. That check exists but it doesn't affect the redacting.